### PR TITLE
JIRA-84: Introduce a new API for handling the authentication

### PR DESCRIPTION
**Pull request by Josue-T**
Implement https://jira.xwiki.org/browse/JIRA-83 and https://jira.xwiki.org/browse/JIRA-84

Note that the unit test need some more love to increase the coverage, but I would like firstly a review to see if it's mostly good or if some change might be needed.

Also the Admin UI need to be adapted. This will be done as a second step.

### CODE REVIEW

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                        basicAuthObj.setStringValue(PROPERTY_ID, serverId);
+                        basicAuthObj.setStringValue(PROPERTY_USERNAME, username);
+                        basicAuthObj.setStringValue(PROPERTY_PASSWORD, password);
+                    }
+                }
+                // TODO set correct user for save...
+                xwiki.saveDocument(basicAuthDoc, DOC_SAVE_MESSAGE, context);
+                xwiki.saveDocument(jiraServerDoc, DOC_SAVE_MESSAGE, context);
+                // Juste create a single document to save the fact that we run the migration
+                xwiki.saveDocument(xwiki.getDocument(migrationDocumentRef, context).clone(), context);
```
Comment (Josue-T) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:133]: Currently the document is marked as saved by XWikiGuest, what would be the recommended way for this ?

File: jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.hc.core5.http.HttpHost;
+import org.xwiki.component.annotation.Role;
+
+import com.atlassian.jira.rest.client.api.AuthenticationHandler;
+
+/**
+ * JIRA authenticator, to send authenticated request to the target JIRA server.
+ *
+ * @version $Id$
+ * @since 11.3.0
```
Comment (surli) [inline jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java:33]: Wrong since: that would be a 10.3.0 or a 11.0.0 depending if we're going to a major release or not.

File: jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+ */
+@Role
+public interface JIRAAuthenticator
+{
+    /**
+     * @return the JiraRestClient instance configured by the provider.
```
Comment (surli) [inline jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java:39]: wrong doc?

File: jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+     */
+    AuthenticationHandler getRestClientAuthenticationHandler();
+
+    /**
+     * Authenticate in HttpClient.
```
Comment (surli) [inline jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java:44]: Doesn't provide any info...

File: jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAServer.java [hunk 5]
```diff
@@ -30,34 +32,35 @@
...
+     * Constructor for JIRA server which will send the requests with authentication.
      *
      * @param url see {@link #getURL()}
-     * @param username see {@link #getUsername()}
-     * @param password see {@link #getPassword()}
+     * @param id see {@link #getId()}
+     * @param jiraAuthenticator see {@link #getJiraAuthenticator()}
      */
-    public JIRAServer(String url, String username, String password)
+    public JIRAServer(String url, String id, JIRAAuthenticator jiraAuthenticator)
```
Comment (surli) [inline jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAServer.java:60]: You're breaking an API here, is it on purpose? Can't you keep the API and delegate to an authenticator?

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/BasicAuthJIRAAuthenticatorFactory.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+                .getDocument(BASIC_AUTH_CONFIG_REFERENCE, context);
+            Optional<BaseObject> authObj = doc.getXObjects(BASIC_AUTH_DATA_CLASS_REFERENCE)
+                .stream().filter(x -> StringUtils.equals(serverId, (x.getStringValue(CONFIG_ID_FIELD))))
+                .findFirst();
+            if (authObj.isPresent()) {
+                String username = authObj.get().getStringValue("username");
+                String password = authObj.get().getStringValue("password");
+                return new BasicAuthJIRAAuthenticator(username, password);
+            } else {
+                throw new RuntimeException("Can't find Basic auth config for server ID: " + serverId);
```
[the remark this answers is not in the document]
Comment (Josue-T) [reply jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/BasicAuthJIRAAuthenticatorFactory.java:93]: We can, also use an `Optional`, but note that this case is not really expected. I mean, if the JIRAServer is configured to have a Basic Auth  authenticator but then, if we can't find the basic auth object with the credentials there are a problem.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java [hunk 7]
```diff
@@ -107,13 +120,19 @@ protected Object getBaseProperty(String propertyName, boolean text) throws XWiki
...
+                    jiraServer = new JIRAServer(urlProperty.getValue(), idProperty.getValue());
                 } else {
-                    jiraServer = new JIRAServer(urlProperty.getValue());
+                    try {
+                        JIRAuthenticatorFactory jirAuthenticatorFactory = componentManagerProvider.get()
+                            .getInstance(JIRAuthenticatorFactory.class, authType);
+                        jiraServer = new JIRAServer(urlProperty.getValue(), idProperty.getValue(),
+                            jirAuthenticatorFactory.get(idProperty.getValue()));
+                    } catch (ComponentLookupException e) {
+                        logger.error("Can't find JIRAuthenticatorFactory with name [{}]", authType, e);
```
Comment (surli) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java:133]: I think I'd do it differently: here it's easy for an admin to entirely miss the fact that a configured server instance doesn't work because of a bad authenticator name without having any clue in the UI. Also you're mixing the fact that the component hint isn't right and the fact that the component cannot be properly initialized. 
So I'd use a if to check if the hint exists first, if it's not the case I'd call the JIRAServer constructor with null, and add some checks in JIRAServer to test if it's null and be able to display that in the config UI. And then if the hint exists but there's a problem to instantiate it I'd throw an exception.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+import org.xwiki.user.internal.document.DocumentUserReference;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+
+/**
+ * Listener to handle migration of authentication in JIRA configuration.
```
Comment (surli) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:48]: That's not a listener...

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java [hunk 9]
```diff
@@ -131,6 +150,7 @@ private boolean isPropertyEmpty(BaseProperty property)
 
     private List<BaseObject> getJIRAServerBaseObjects() throws XWikiException
     {
+        configurationMigrator.migrate();
```
[the remark this answers is not in the document]
Comment (Josue-T) [reply jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java:153]: I totally agree that the current state it not nice, initially I wrote this with a listener with the event `ExtensionUpgradedEvent("org.xwiki.contrib.jira:jira-config")` but it wasn't working at all and I didn't really understood why.  So I wrote this until I find a better way to do this.

File: jira-config/jira-config-ui/src/main/resources/JIRA/JIRAConfig.xml [hunk 10]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="UTF-8"?>
+<?xml version="1.1" encoding="UTF-8"?>
```
Comment (surli) [inline jira-config/jira-config-ui/src/main/resources/JIRA/JIRAConfig.xml:1]: Should be removed from the changes probably.

File: jira-macro/jira-macro-common/pom.xml [hunk 11]
```diff
@@ -33,16 +33,17 @@
...
     <dependency>
       <groupId>org.jdom</groupId>
       <artifactId>jdom2</artifactId>
     </dependency>
     <dependency>
-      <groupId>org.apache.httpcomponents</groupId>
-      <artifactId>httpclient</artifactId>
+      <groupId>org.apache.httpcomponents.client5</groupId>
+      <artifactId>httpclient5</artifactId>
+      <version>5.2</version>
```
Comment (surli) [inline jira-macro/jira-macro-common/pom.xml:46]: Not sure why you need this upgrade, but I would put the version directly as a property of the root pom to ease future upgrades.

File: jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAServer.java (hunk 5 above)
Comment (Josue-T) [reply jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAServer.java:60]: I agree that it's a breaking change, but from our quick search this API don't seem used a lot, so it's why we considered as it's ok to break it. But if you prefer to keep it I can do a workaround to keep it.

File: jira-macro/jira-macro-platform/src/main/java/org/xwiki/contrib/jira/config/internal/AsyncJIRAMacro.java [hunk 12]
```diff
@@ -80,7 +80,11 @@ public List<Block> execute(JIRAMacroParameters parameters, String content, Macro
 
         // Take into account the current wiki since the jira config document can be located in each wiki (without this
         // all config document will be taken from the main wiki).
-        configuration.setContextEntries(Collections.singleton(XWikiContextContextStore.PROP_WIKI));
+        configuration.setContextEntries(Set.of(
+            XWikiContextContextStore.PROP_WIKI,
+            XWikiContextContextStore.PROP_USER,
+            XWikiContextContextStore.PROP_REQUEST_URL
+        ));
```
Comment (surli) [inline jira-macro/jira-macro-platform/src/main/java/org/xwiki/contrib/jira/config/internal/AsyncJIRAMacro.java:87]: You forgot to update the comment to explain why you need all that now.

File: jira-macro/jira-macro-common/src/main/java/org/xwiki/contrib/jira/macro/internal/JIRAMacroTransformationManager.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+        // Add additional transformation from others components
+        try {
+            Type t = new DefaultParameterizedType(null, JIRAMacroTransformation.class, parameters.getClass());
+            List<JIRAMacroTransformation<P>> jiraMacroTransformations =
+                componentManagerProvider.get().getInstanceList(t);
+            for (JIRAMacroTransformation<P> transformation : jiraMacroTransformations) {
+                resultingBlocks = transformation.transform(resultingBlocks, parameters, context, jiraServer, macroName);
+            }
+        } catch (ComponentLookupException e) {
+            throw new RuntimeException(e);
```
Comment (surli) [inline jira-macro/jira-macro-common/src/main/java/org/xwiki/contrib/jira/macro/internal/JIRAMacroTransformationManager.java:77]: Not good, you need specific exception.

File: jira-macro/jira-macro-common/pom.xml (hunk 11 above)
Comment (Josue-T) [reply jira-macro/jira-macro-common/pom.xml:46]: The main reason to upgrade it, is that the Httpclient will be used by the external API of the `JIRAAuthenticator` and this API will be used by an external extension so after this it might be more complicated to upgrade it.

File: jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/bidimensionalgrid/JIRABiDimensionalGridChartMacro.java [hunk 14]
```diff
@@ -107,7 +115,8 @@ public List<Block> execute(JIRABiDimensionalGridChartMacroParameter parameters,
         }
 
         TableBlock tableBlock = new TableBlock(tableRows);
-        return List.of(tableBlock);
+        return jiraMacroTransformationManager.transform(List.of(tableBlock), parameters, context,
+            jiraServerResolver.resolve(parameters), JIRABiDimensionalGridChartMacro.MACRO_NAME);
```
Comment (surli) [inline jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/bidimensionalgrid/JIRABiDimensionalGridChartMacro.java:119]: Sounds like it should be part of AbstractMacro no?

File: jira-macro/jira-macro-common/src/main/java/org/xwiki/contrib/jira/macro/JIRAMacroTransformation.java [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Interface used to apply a transformation after the rendering of the jira, jiraCount an jiraChart macros.
+ * @param <P> the type of Macro parameter.
+ *
+ * @version $Id$
+ * @since 10.3.0
+ */
+@Role
+public interface JIRAMacroTransformation<P>
```
Comment (surli) [inline jira-macro/jira-macro-common/src/main/java/org/xwiki/contrib/jira/macro/JIRAMacroTransformation.java:37]: I don't see any implem of it here, did I miss it?

File: jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAServer.java (hunk 5 above)
Comment (surli) [reply jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAServer.java:60]: Well, let's see if you can deprecate the original API while keeping it: if it's making things too complicated we'll remove it.

File: jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java (hunk 2 above)
Comment (surli) [reply jira-config/jira-config-api/src/main/java/org/xwiki/contrib/jira/config/JIRAAuthenticator.java:33]: Given your changes and the possible breaking APIs, it feels like a next major version so 11.0.0

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/BasicAuthJIRAAuthenticatorFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (surli) [reply jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/BasicAuthJIRAAuthenticatorFactory.java:93]: So the question is: should we consider that in normal condition the factory should always be able to build an instance? If it's true then I agree to keep current signature, and to throw an exception when the factory cannot build the instance. If it's not true and we can imagine cases where the factory in normal condition cannot build the instance, then it should return an optional. 

Note that the signature should still be changed to be able to throw a more typed exception than RuntimeException.

File: jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/bidimensionalgrid/JIRABiDimensionalGridChartMacro.java (hunk 14 above)
Comment (surli) [reply jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/bidimensionalgrid/JIRABiDimensionalGridChartMacro.java:119]: Yes, if it could avoid making each time the same call in all implems, and to inject same component.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Public constructor.
+     */
+    public JIRAConfigurationMigrator()
+    {
+        super(HINT, List.of(new ApplicationReadyEvent(), new JobFinishedEvent("install")));
```
Comment (tmortagne) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:100]: Note that ApplicationReadyEvent only cover main wiki use case, you might want to also listen to WikiReadyEvent.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        migrate();
```
Comment (tmortagne) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:106]: When you receive JobFinishedEvent the context wiki will always be the main wiki, but the upgrade might have been actually been done on a subwiki and I assume you need to do the migration in a subwiki in that case.

Comment (surli) [reply jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:106]: IMO you need a new method used as an oracle to know if migrate should be called or not: right now this listener will be called each time the wiki is restarted and will call migrate() which doesn't perform any check.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+        throws XWikiException
+    {
+        XWikiContext context = contextProvider.get();
+        XWiki xwiki = context.getWiki();
+        DocumentReference migrationDocumentRef =
+            new DocumentReference(MIGRATION_STATUS_DOCUMENT_REFERENCE, new WikiReference(context.getWikiId()));
+
+        XWikiDocument migrationStateDoc = xwiki.getDocument(migrationDocumentRef, context).clone();
+        UserReference userRef = new DocumentUserReference(
+            new DocumentReference("xwiki", "XWiki", "superadmin"), true);
```
Comment (surli) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:177]: Better use `SuperUserAdminReference.INSTANCE`

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+            new DocumentReference(MIGRATION_STATUS_DOCUMENT_REFERENCE, new WikiReference(context.getWikiId()));
+
+        try {
+            if (!xwiki.exists(migrationDocumentRef, context)) {
+                logger.info("Running migration of JIRA configuration");
+
+                XWikiDocument basicAuthDoc =
+                    xwiki.getDocument(BasicAuthJIRAAuthenticatorFactory.BASIC_AUTH_CONFIG_REFERENCE, context).clone();
+                XWikiDocument jiraServerDoc =
+                    xwiki.getDocument(new LocalDocumentReference(JIRA, "JIRAConfig"), context).clone();
```
Comment (surli) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:126]: why do you keep cloning your docs?

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java [hunk 20]
```diff
@@ -124,6 +132,36 @@ protected Object getBaseProperty(String propertyName, boolean text) throws XWiki
...
+    {
+        ComponentManager componentManager = componentManagerProvider.get();
+
+        JIRAServer jiraServer;
+        if (StringUtils.isEmpty(authType) || "noAuth".equals(authType)) {
+            jiraServer = new JIRAServer(urlProperty.getValue(), idProperty.getValue());
+        } else {
+            if (!componentManager.hasComponent(JIRAuthenticatorFactory.class, authType)) {
+                logger.error("Can't find JIRAuthenticatorFactory with name [{}]", authType);
+                jiraServer = new JIRAServer(urlProperty.getValue(), idProperty.getValue());
```
Comment (surli) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java:145]: Why not returning null here?

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java [hunk 21]
```diff
@@ -124,6 +132,36 @@ protected Object getBaseProperty(String propertyName, boolean text) throws XWiki
...
+            } else {
+                try {
+                    JIRAuthenticatorFactory jirAuthenticatorFactory = componentManagerProvider.get()
+                        .getInstance(JIRAuthenticatorFactory.class, authType);
+                    jiraServer = new JIRAServer(urlProperty.getValue(), idProperty.getValue(),
+                        jirAuthenticatorFactory.get(idProperty.getValue()));
+                } catch (ComponentLookupException e) {
+                    logger.error("Can't create JIRAuthenticatorFactory for authentication type [{}]", authType,
+                        e);
+                    return null;
```
Comment (surli) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java:155]: would be cleaner to initialize jiraServer at null so you'd always return null in the last statement in case of problem and you wouldn't have to do anything else than logging an error if the hint is not found in CM

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java (hunk 19 above)
Comment (Josue-T) [reply jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:126]: Because I suppose we should always clone it if we make modification on.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java (hunk 20 above)
Comment (Josue-T) [reply jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigClassDocumentConfigurationSource.java]: Done

File: jira-config/jira-config-api/pom.xml [hunk 22]
```diff
@@ -51,5 +51,41 @@
       <version>${commons.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.httpcomponents.client5</groupId>
+      <artifactId>httpclient5</artifactId>
+      <version>5.2</version>
+    </dependency>
+    <dependency>
```
Comment (tmortagne) [inline jira-config/jira-config-api/pom.xml:59]: Since several different modules depend on this I would at the very least put a Maven property to store the version in the root pom of the jira project.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+ */
+@Singleton
+@Component
+@Named(JIRAConfigurationMigrator.HINT)
+public class JIRAConfigurationMigrator extends AbstractEventListener
+{
+    /**
+     * Main name of the listener.
+     */
+    public static final String HINT = "JIRAConfigMigrationListener";
```
Comment (tmortagne) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:65]: It's not a technical problem in any way, but it feels like it initially was the name of the class, and then you change the class name.

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java [hunk 24]
```diff
@@ -0,0 +1,188 @@
...
+
+    @Inject
+    private Logger logger;
+
+    /**
+     * Public constructor.
+     */
+    public JIRAConfigurationMigrator()
+    {
+        super(HINT, List.of(new ApplicationReadyEvent(), new WikiReadyEvent(), new JobFinishedEvent("install")));
```
Comment (tmortagne) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/JIRAConfigurationMigrator.java:93]: You don't need to go through List.of, there is a `AbstractEventListener(String name, Event... events)`.

File: jira-config/jira-config-api/pom.xml (hunk 22 above)
Comment (tmortagne) [reply jira-config/jira-config-api/pom.xml:58]: Not sure what you mean here, I don't remember suggesting to remove any exclusion. As I said, one benefit of setting up a dependency management is that you can put the exclusions in this central place, and it will apply to all modules where you have this dependency.

Comment (Josue-T) [reply jira-config/jira-config-api/pom.xml:58]: Didn't full understood what you was thinking but now it should be good.

File: jira-config/jira-config-platform/pom.xml [hunk 25]
```diff
@@ -52,5 +52,10 @@
       <artifactId>xwiki-platform-configuration-default</artifactId>
       <version>${platform.version}</version>
     </dependency>
+    <dependency>
+      <groupId>org.xwiki.platform</groupId>
+      <artifactId>xwiki-platform-user-default</artifactId>
```
Comment (surli) [inline jira-config/jira-config-platform/pom.xml:57]: I don't think you need it anymore.

File: jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/AbstractJIRAChartMacro.java [hunk 26]
```diff
@@ -85,4 +101,22 @@ protected String getJSONData(T parameters, Class<U> dataSourceClass) throws Macr
...
+
+    /**
+     * Apply macro transformation on the macro result.
+     *
+     * @param parameters the macro parameter, depending on the type of the parameter.
+     * @param context the macro execution context.
+     * @param macroBlock the input macro block to apply the transformation.
+     * @param macroName the macro name which was called.
+     * @return the list of block after the transformation was called.
+     * @throws MacroExecutionException in case there are a {@link JIRAServer} resolution issue
```
Comment (surli) [inline jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/AbstractJIRAChartMacro.java:113]: You should add a @since in new methods javadoc.

File: jira-macro/jira-macro-common/src/main/java/org/xwiki/contrib/jira/macro/internal/HTTPJIRAFetcher.java [hunk 27]
```diff
@@ -63,10 +59,14 @@ public class HTTPJIRAFetcher
 {
     private static final ErrorMessageExtractor EXTRACTOR = new ErrorMessageExtractor();
 
+    @Inject
+    @Named("context")
+    private Provider<ComponentManager> componentManagerProvider;
```
[the remark this answers is not in the document]
Comment (surli) [reply jira-macro/jira-macro-common/src/main/java/org/xwiki/contrib/jira/macro/internal/HTTPJIRAFetcher.java:64]: wdym? It's still there AFAICS...

File: jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/BasicAuthJIRAAuthenticatorFactory.java [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+    {
+        XWikiContext context = contextProvider.get();
+        XWikiDocument doc;
+        try {
+            doc = context.getWiki().getDocument(BASIC_AUTH_CONFIG_REFERENCE, context);
+        } catch (XWikiException e) {
+            throw new JIRAAuthenticatorException("Can't get JIRA Basic auth configuration document", e);
+        }
+        Optional<BaseObject> authObj = doc.getXObjects(BASIC_AUTH_DATA_CLASS_REFERENCE)
+            .stream().filter(x -> StringUtils.equals(serverId, (x.getStringValue(CONFIG_ID_FIELD))))
```
Comment (surli) [inline jira-config/jira-config-platform/src/main/java/org/xwiki/contrib/jira/config/internal/BasicAuthJIRAAuthenticatorFactory.java:89]: You forgot to check if x != null here.

File: jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/AbstractJIRAChartMacro.java [hunk 29]
```diff
@@ -69,6 +85,9 @@ public boolean supportsInlineMode()
 
     /**
      * Fetch the data, convert them, and serialize them to JSON to be used in ChartJS macro.
+     *
+     * @since 11.0.0
```
Comment (surli) [inline jira-macro/jira-macro-charts/src/main/java/org/xwiki/contrib/jira/charts/internal/AbstractJIRAChartMacro.java:89]: Not added at the right place :)

File: pom.xml [hunk 30]
```diff
@@ -74,12 +74,52 @@
...
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-core</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-slf4j-impl</artifactId>
+          </exclusion>
+          <exclusion>
```
Comment (tmortagne) [inline pom.xml:101]: It's not very clear why guava is excluded, since it's not a forbidden dependency (same thing for error_prone_annotations, checker-qual, jersey-client and ersey-media-json-jettison).

Comment (tmortagne) [reply pom.xml:101]: You have to keep in mind that 90% of the time, exclusions always hide a real problem. Every exclusion you add an if a something that either won't be there at all or with an older version that what as indicated when this extension is installed in XWiki 15.10. There is a reason we have those checks, it's not just to annoy contrib extension developers.

For example:

>         +-com.google.guava:guava:31.1-jre (managed) <-- com.google.guava:guava:32.0.1-jre

Are you sure that jira-rest library does not really need something that was introduced in guava 32 ?

> For jersey the issue is that it require jakarta which is banned in XWiki-platform 15.10 (from what I understood).

OK, but then it means that if it does need jersey-client (and I assume it does, as otherwise I don't see why it would indicate it as dependency), it simply won't work in XWiki 15.10 (because that dependency won't be there).

>     +-com.google.errorprone:error_prone_annotations:2.16 (managed) <-- com.google.errorprone:error_prone_annotations:2.18.0

For error_prone_annotations you are lucky: it should be fine because it's just there to declare annotations, which are generally only used at build time. But again, it's important to explain in the pom why each exclusion is there and why it's OK.

> I would be pleased to know what would be the recommended way to solve this.

Unfortunately, often there is simply no fix: it just means that you are using something which is not compatible with your target version of XWiki (warning about that risk is the whole point of that check).

Comment (tmortagne) [reply pom.xml:101]: I cannot really answer these question since I don't know this part of the jira project. I'm only the annoying reviewer reporting the fishy exclusions.

Comment (Josue-T) [reply pom.xml:101]: ah my bad, I didn't express myself clearly. I was meaning banned, not excluded... 🫣 so I edited my message... anyway my question is why jersey (which seem to depends on Jakarta) is compatible with xwiki 14.10 but not 15.10.
